### PR TITLE
feat: mad-navigation

### PR DESCRIPTION
**Pull request by tormodAase**
First version of mad-navigation

### CODE REVIEW

File: packages/navigation/src/native-stack/createNativeStackNavigator.tsx
Comment (HaakonSvane) [inline packages/navigation/src/native-stack/createNativeStackNavigator.tsx:1]: If this file is copied from the RNN source code (which I seem to remember you told me), perhaps we should inform the users of the source of it in a top level comment or something similar. This makes it easier to fix potential issues in the future as well

File: packages/navigation/src/native-stack/types.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export type NativeStackNavigatorProps = DefaultNavigatorOptions<
+    ParamListBase,
+    StackNavigationState<ParamListBase>,
+    MadNativeStackNavigationOptions,
+    NativeStackNavigationEventMap
+> &
+    StackRouterOptions &
+    NativeStackNavigationConfig;
+
+// eslint-disable-next-line @typescript-eslint/ban-types -- those fuckers don't export this useless type. If they export it in the future, please import it from native-stack and remove it from here. Yes it is intended to be empty. https://github.com/react-navigation/react-navigation/blob/main/packages/native-stack/src/types.tsx
```
Comment (HaakonSvane) [inline packages/navigation/src/native-stack/types.ts:22]: 😂

File: packages/navigation/jest.config.d.cts
Comment (HaakonSvane) [inline packages/navigation/jest.config.d.cts:1]: I haven't seen this file before, similar to the tsup config file. I think it was generated as part of a build going bananas. Should perhaps be removed

File: packages/navigation/tsup.config.ts
Comment (HaakonSvane) [inline packages/navigation/tsup.config.ts:1]: This file seems to be copied from mad components, and therefore contains unnecessary code for attempting to load fonts into the dist folder at build.
Please remove all mentions of font loading. You can look to the skia draw config for inspiration on a more straight-forward build config.

**Review by HaakonSvane — APPROVED** [review]
Truly amazing. Great to see core starting to take shape. Super awesome work.

File: packages/navigation/src/_internal/types.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+export type MadNavigationOptions = { environmentBannerShown?: boolean };
+
+export type ScreenOptions<Options extends MadNavigationOptions> =
+    | Options
+    | ((props: { route: FinalRouteProp; navigation: unknown }) => Options)
+    | undefined;
+
+export type FinalRouteProp = RouteProp<ParamListBase, string>;
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/src/_internal/types.ts:22]: There's nothing "mad" about this prop, it's just a common type that I decided to simplify. If you have any other suggestions though let me know

File: packages/navigation/README.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            options={{
+                // or here
+                environmentBannerShown: false,
+            }}
+        />
+    </Stack.Group>
+</Stack.Navigator>
+```
+
+_CAUTION: Environmentbanners will not work properly if `headerLargeTitle` is set to true, and
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/README.md]: all typos should be fixed

File: packages/navigation/src/native-stack/createNativeStackNavigator.tsx
Comment (tormodAase) [reply packages/navigation/src/native-stack/createNativeStackNavigator.tsx:1]: Good idea! added links

File: packages/navigation/tsup.config.ts
Comment (tormodAase) [reply packages/navigation/tsup.config.ts:1]: Think we have to look at this together

File: packages/navigation/jest.config.d.cts
Comment (tormodAase) [reply packages/navigation/jest.config.d.cts:1]: removed jest.config.d.cts and tsup.config.d.ts

File: packages/navigation/package.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    },
+    "license": "MIT",
+    "bugs": {
+        "url": "https://github.com/equinor/mad/issues"
+    },
+    "homepage": "https://github.com/equinor/mad/",
+    "main": "./dist/index.js",
+    "types": "./dist/index.d.ts",
+    "peerDependencies": {
+        "react": ">=16.8",
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/package.json:28]: This should be fixed, though the PR linting is complaining about something

File: packages/navigation/src/_internal/getFinalScreenOptions.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { FinalRouteProp, MadNavigationOptions, ScreenOptions } from "./types";
+
+/**
+ * Screenoptions can be an object, a function, or undefined. This function calls it with the right properties if it is a function, resulting in an object or undefined
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/src/_internal/getFinalScreenOptions.ts:4]: Fixed

File: packages/navigation/package.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    "type": "module",
+    "scripts": {
+        "build": "tsup-node",
+        "test": "jest",
+        "dev": "tsup-node --watch --clean=false",
+        "lint": "eslint src/"
+    },
+    "keywords": [
+        "react native",
+        "ui",
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/package.json]: Keywords fixed

File: packages/navigation/src/_internal/getFinalScreenOptions.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+ * @param screenOptions
+ * @param route
+ * @param navigation
+ * @returns screenOptions
+ */
+export const getFinalScreenOptions = <T extends MadNavigationOptions>(
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/src/_internal/getFinalScreenOptions.ts:10]: Renamed and modified some parts of the code to make the flow better

File: packages/navigation/src/_internal/getFinalOptions.ts
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/src/_internal/getFinalOptions.ts:1]: docs should be updated!

File: packages/navigation/src/_internal/getFinalScreenOptions.ts
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/src/_internal/getFinalScreenOptions.ts:1]: docs should be updated!

File: packages/navigation/src/_internal/shouldDisplayEnvironmentBanner.ts
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/src/_internal/shouldDisplayEnvironmentBanner.ts:1]: docs should be updated!

File: packages/navigation/src/_internal/getFinalOptions.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { MadBaseOptions, MadDescriptorBase } from "./types";
+
+/**
+ * Given the descriptor, and the screenOptions, should return the final options we have to work with
+ * @param descriptor
+ * @param screenOptions
+ */
+export const getFinalOptions = (
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/src/_internal/getFinalOptions.ts:8]: Great idea, i've renamed some of the functions

File: packages/navigation/README.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+You also have to install the navigators you want to use, e.g `@react-navigation/bottom-tabs` and
+`@react-navigation/native-stack`
+
+If you want environment banners to display, you also need to install `@equinor/mad-components`
+version 0.5.0 or higher (unless you use `mad-core`)
+
+### Usage
+
+Follow [React Navigation’s documentation](https://reactnavigation.org/docs/getting-started/). When
+you get to the point where you are creating a navigator, import `createXNavigator` from
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply packages/navigation/README.md]: Updated

**Review by HaakonSvane — APPROVED** [review]
👍

**Pull request closed without merging**